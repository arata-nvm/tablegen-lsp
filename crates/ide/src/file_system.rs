use std::collections::{HashMap, VecDeque};
use std::path::{Path, PathBuf};

use ecow::EcoString;
use syntax::ast::AstNode;
use syntax::parser::{TextRange, TextSize};
use syntax::{SyntaxNode, SyntaxNodePtr, ast};

use crate::db::{Db, parse};

#[derive(Debug, Clone, Copy, Ord, PartialOrd, Eq, PartialEq, Hash)]
pub struct FileId(pub u32);

#[derive(Debug, Clone, Copy, Eq, PartialEq, Hash)]
pub struct FilePosition {
    pub file: FileId,
    pub position: TextSize,
}

impl FilePosition {
    pub fn new(file: FileId, position: TextSize) -> Self {
        Self { file, position }
    }
}

#[derive(Debug, Clone, Copy, Eq, PartialEq, Hash)]
pub struct FileRange {
    pub file: FileId,
    pub range: TextRange,
}

impl FileRange {
    pub fn new(file: FileId, range: TextRange) -> Self {
        Self { file, range }
    }
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct FilePath(pub PathBuf);

impl FilePath {
    pub fn join(&self, path: impl AsRef<Path>) -> FilePath {
        FilePath(self.0.join(path))
    }

    pub fn parent(&self) -> Option<FilePath> {
        self.0.parent().map(FilePath::from)
    }
}

impl From<&Path> for FilePath {
    fn from(value: &Path) -> Self {
        Self(value.to_path_buf())
    }
}

#[derive(Debug, Default)]
pub struct FileSet {
    path_to_id: HashMap<FilePath, FileId>,
    id_to_path: HashMap<FileId, FilePath>,
}

impl FileSet {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn insert(&mut self, file_id: FileId, path: FilePath) {
        self.path_to_id.insert(path.clone(), file_id);
        self.id_to_path.insert(file_id, path.clone());
    }

    pub fn remove(&mut self, file_id: &FileId) {
        if let Some(path) = self.id_to_path.remove(file_id) {
            self.path_to_id.remove(&path);
        }
    }

    pub fn contains(&self, file_id: &FileId) -> bool {
        self.id_to_path.contains_key(file_id)
    }

    pub fn file_for_path(&self, path: &FilePath) -> Option<FileId> {
        self.path_to_id.get(path).copied()
    }

    pub fn path_for_file(&self, file_id: &FileId) -> &FilePath {
        &self.id_to_path[file_id]
    }

    pub fn iter_files(&self) -> impl Iterator<Item = FileId> + '_ {
        self.id_to_path.keys().copied()
    }
}

#[derive(Debug, Clone, Copy, Ord, PartialOrd, Eq, PartialEq, Hash)]
pub struct SourceUnitId(FileId);

impl SourceUnitId {
    pub fn from_root_file(root_file: FileId) -> Self {
        Self(root_file)
    }
}

#[derive(Debug)]
pub struct SourceUnit {
    root: FileId,
    includes: HashMap<FileId, IncludeMap>,
}

impl SourceUnit {
    pub fn root(&self) -> FileId {
        self.root
    }

    pub fn iter_files(&self) -> impl Iterator<Item = FileId> + '_ {
        self.includes.keys().copied()
    }

    pub fn include_map(&self, file_id: &FileId) -> Option<&IncludeMap> {
        self.includes.get(file_id)
    }
}

pub trait FileSystem {
    fn assign_or_get_file_id(&mut self, path: FilePath) -> FileId;

    fn path_for_file(&self, file_id: &FileId) -> &FilePath;

    fn read_content(&self, file_path: &FilePath) -> Option<String>;
}

pub fn collect_sources<FS: FileSystem>(
    db: &mut dyn Db,
    fs: &mut FS,
    root: FileId,
    include_dirs: &[FilePath],
) -> SourceUnit {
    let mut includes = HashMap::new();

    let mut queue = VecDeque::new();
    queue.push_back(root);
    while let Some(file_id) = queue.pop_front() {
        if includes.contains_key(&file_id) {
            continue;
        }

        let parse = parse(db, file_id);

        let file_path = fs.path_for_file(&file_id);
        let file_dir = file_path.parent().expect("file dir not found");

        let mut include_dir_list = vec![file_dir];
        include_dir_list.extend(include_dirs.iter().cloned());

        let mut include_map = HashMap::new();
        for (include_id, include_path) in list_includes(parse.syntax_node()) {
            let Some(resolved_file_id) =
                resolve_include_file(db, fs, include_path, &include_dir_list)
            else {
                continue;
            };

            include_map.insert(include_id, resolved_file_id);
            queue.push_back(resolved_file_id);
        }
        includes.insert(file_id, include_map);
    }

    SourceUnit { root, includes }
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct IncludeId(pub SyntaxNodePtr);

pub type IncludeMap = HashMap<IncludeId, FileId>;

fn list_includes(root_node: SyntaxNode) -> Vec<(IncludeId, EcoString)> {
    (|| -> Option<_> {
        let source_file = ast::SourceFile::cast(root_node)?;
        let stmt_list = source_file.statement_list()?;
        let include_paths = stmt_list
            .statements()
            .filter_map(|stmt| match stmt {
                ast::Statement::Include(include) => {
                    let id = IncludeId(SyntaxNodePtr::new(include.syntax()));
                    let path = include.path()?.value();
                    Some((id, path))
                }
                _ => None,
            })
            .collect();
        Some(include_paths)
    })()
    .unwrap_or_default()
}

fn resolve_include_file<FS: FileSystem>(
    db: &mut dyn Db,
    fs: &mut FS,
    include_path: EcoString,
    include_dir_list: &[FilePath],
) -> Option<FileId> {
    for include_dir in include_dir_list {
        let candidate_file_path = include_dir.join(include_path.as_str());
        if let Some(file_content) = fs.read_content(&candidate_file_path) {
            let file_id = fs.assign_or_get_file_id(candidate_file_path.clone());
            db.set_file_content(file_id, file_content.as_str());
            return Some(file_id);
        }
    }
    None
}
