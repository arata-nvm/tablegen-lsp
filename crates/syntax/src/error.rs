use ecow::EcoString;
use rowan::TextRange;
use std::fmt;

#[derive(Debug, Clone)]
pub struct SyntaxError {
    pub range: TextRange,
    pub message: EcoString,
}

impl SyntaxError {
    pub fn new(range: TextRange, message: impl Into<EcoString>) -> Self {
        Self {
            range,
            message: message.into(),
        }
    }
}

impl fmt::Display for SyntaxError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{:?}:{}", self.range, self.message)
    }
}
