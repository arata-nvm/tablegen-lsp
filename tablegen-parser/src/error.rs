use std::fmt;

use ecow::EcoString;

pub type Position = usize;
pub type Range = std::ops::Range<Position>;

#[derive(Debug, Clone)]
pub struct SyntaxError {
    pub range: Range,
    pub message: EcoString,
}

impl SyntaxError {
    pub fn new(range: Range, message: impl Into<EcoString>) -> Self {
        Self {
            range,
            message: message.into(),
        }
    }
}

impl fmt::Display for SyntaxError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "{}:{}:{}",
            self.range.start, self.range.end, self.message
        )
    }
}
