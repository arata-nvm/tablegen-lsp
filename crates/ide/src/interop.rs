use std::str::FromStr;
use std::{collections::HashMap, str::Utf8Error};

use ecow::EcoString;
pub use tblgen::diagnostic::DiagnosticKind;
use tblgen::error::SourceLoc;
use tblgen::{RecordKeeper, TableGenParser};

use crate::file_system::{FilePath, FilePosition, FileSystem};

#[derive(Debug, Clone, Eq, PartialEq)]
pub struct TblgenParseResult {
    pub diagnostics: Vec<TblgenDiagnostic>,
    pub symbol_table: TblgenSymbolTable,
}

#[derive(Debug, Clone, Eq, PartialEq)]
pub struct TblgenDiagnostic {
    pub kind: tblgen::diagnostic::DiagnosticKind,
    pub message: String,
    pub filename: String,
    pub line: i32,
    pub column: i32,
}

pub fn parse_source_unit_with_tblgen(
    root_file: &FilePath,
    include_dirs: &[FilePath],
    fs: &impl FileSystem,
) -> Result<TblgenParseResult, Utf8Error> {
    let mut parser = TableGenParser::new().add_source_file(root_file.to_str());
    if let Some(parent) = root_file.parent() {
        parser = parser.add_include_directory(parent.to_str());
    }
    for include_dir in include_dirs {
        parser = parser.add_include_directory(include_dir.to_str());
    }
    let result = parser.parse();

    Ok(TblgenParseResult {
        diagnostics: result
            .diagnostics
            .into_iter()
            .map(convert_diagnostic)
            .collect::<Result<_, _>>()?,
        symbol_table: TblgenSymbolTable::with_record_keeper(&result.record_keeper, fs),
    })
}

fn convert_diagnostic(
    diag: tblgen::diagnostic::Diagnostic<'_>,
) -> Result<TblgenDiagnostic, Utf8Error> {
    let message = diag.message();
    let filename = diag.filename();
    Ok(TblgenDiagnostic {
        kind: diag.kind(),
        message: message.as_str()?.to_string(),
        filename: filename.as_str()?.to_string(),
        line: diag.line() - 1,
        column: diag.column(),
    })
}

#[derive(Debug, Default, Clone, Eq, PartialEq)]
pub struct TblgenSymbolTable {
    pub defs: HashMap<FilePosition, Vec<TblgenDef>>,
}

#[derive(Debug, Default, Clone, Eq, PartialEq)]
pub struct TblgenDef {
    pub name: EcoString,
    pub direct_super_classes: Vec<EcoString>,
}

impl TblgenSymbolTable {
    pub fn with_record_keeper(record_keeper: &RecordKeeper, fs: &impl FileSystem) -> Self {
        let mut defs = HashMap::new();
        for (name, record) in record_keeper.defs() {
            let (Ok(name), Some(tblgen_pos)) = (name, record.file_position(record_keeper)) else {
                continue;
            };

            let direct_super_classes = record
                .direct_super_classes()
                .filter_map(|class| class.name().ok())
                .map(EcoString::from)
                .collect::<Vec<_>>();
            let def = TblgenDef {
                name: EcoString::from(name),
                direct_super_classes,
            };

            let tblgen_filepath = tblgen_pos.filepath();
            let file_id = if let Ok(filepath) = tblgen_filepath.as_str()
                && let Ok(filepath) = FilePath::from_str(filepath)
                && let Some(file_id) = fs.file_for_path(&filepath)
            {
                file_id
            } else {
                continue;
            };

            let file_pos = FilePosition::new(file_id, tblgen_pos.pos().into());
            defs.entry(file_pos).or_insert_with(Vec::new).push(def);
        }

        TblgenSymbolTable { defs }
    }

    pub fn get_defs_at(&self, pos: &FilePosition) -> &[TblgenDef] {
        match self.defs.get(pos) {
            Some(defs) => defs,
            None => &[],
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::{file_system::FileSystem, tests};

    #[test]
    fn def() {
        let (_, f) = tests::load_single_file("testdata/def.td");
        let root_file_id = f.root_file();
        let root_file_path = f.path_for_file(&root_file_id);
        let result =
            super::parse_source_unit_with_tblgen(root_file_path, &[], &f).expect("valid code");
        let mut defs = result.symbol_table.defs.into_iter().collect::<Vec<_>>();
        defs.sort_by_key(|(k, _)| k.position);
        insta::assert_debug_snapshot!(defs);
    }
}
