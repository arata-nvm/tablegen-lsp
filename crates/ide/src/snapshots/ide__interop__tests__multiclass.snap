---
source: crates/ide/src/interop.rs
expression: defs
---
[
    (
        FilePosition {
            file: FileId(
                0,
            ),
            position: 169,
        },
        [
            TblgenDef {
                name: "multiclass1_field1",
                direct_super_classes: [
                    "Field",
                ],
            },
            TblgenDef {
                name: "multiclass1_field2",
                direct_super_classes: [
                    "Field",
                ],
            },
        ],
    ),
    (
        FilePosition {
            file: FileId(
                0,
            ),
            position: 201,
        },
        [
            TblgenDef {
                name: "multiclass2_field1",
                direct_super_classes: [
                    "Field",
                ],
            },
            TblgenDef {
                name: "multiclass2_field2",
                direct_super_classes: [
                    "Field",
                ],
            },
        ],
    ),
    (
        FilePosition {
            file: FileId(
                0,
            ),
            position: 236,
        },
        [
            TblgenDef {
                name: "multiclass2_2_field1_field1",
                direct_super_classes: [
                    "Field",
                ],
            },
            TblgenDef {
                name: "multiclass2_2_field1_field2",
                direct_super_classes: [
                    "Field",
                ],
            },
        ],
    ),
]
